rpc.go

```diff
@@ -126,12 +134,61 @@ func (c *client) SendRPC(rpc hrpc.Call) (msg proto.Message, err error) {
 			serverErrorCount++
 			continue // retry
 		case region.NotServingRegionError:
+			if scan, ok := rpc.(*hrpc.Scan); ok {
+				c.scanRpcScanStats(scan, msg, err, true, start, time.Now())
+			}
 			continue // retry
 		}
+		if scan, ok := rpc.(*hrpc.Scan); ok {
+			c.scanRpcScanStats(scan, msg, err, false, start, time.Now())
+		}
 		return msg, err
 	}
 }
 
+func (c *client) scanRpcScanStats(scan *hrpc.Scan, resp proto.Message, err error,
+	retry bool, start, end time.Time) {
+	if scan.ScanStatsHandler() != nil {
+		stats := &hrpc.ScanStats{}
+		// Update the ScanMetrics if they are being tracked. For ScanStats, these ScanMetrics
+		// are collected per call to SendRPC and therefore may not be reflective of the entire
+		// result of the Scan if the results are split across multiple calls to Scanner.Next().
+		if scan.TrackScanMetrics() {
+			scanres, ok := resp.(*pb.ScanResponse)
+			if !ok {
+				c.logger.Error("got non ScanResponse for ScanRequest")
+			} else {
+				if scanres.ScanMetrics != nil {
+					if stats.ScanMetrics == nil {
+						stats.ScanMetrics = make(map[string]int64)
+					}
+					for _, m := range scanres.ScanMetrics.GetMetrics() {
+						stats.ScanMetrics[m.GetName()] += m.GetValue()
+					}
+				}
+			}
+		}
+
+		stats.Table = scan.Table()
+		stats.StartRow = scan.StartRow()
+		stats.EndRow = scan.StopRow()
+		if scan.Region() != nil {
+			stats.RegionID = scan.Region().ID()
+			if scan.Region().Client() != nil {
+				stats.RegionServer = scan.Region().Client().Addr()
+			}
+		}
```

Comment:
A region's client can change in the background, so one call to `scan.Region().Client()` may return non-nil and then later return nil.

The results of `scan.Region()` (though, this one probably can't change in the background) and `scan.Region().Client()` should be saved into variables, ie.

```
if reg := scan.Region(); reg != nil {
    ...
    if c := reg.Client(); c != nil {
        ...
```